### ISSUE TITLE
[filter-effects-2] Support fast simple css blurs

### ISSUE THREAD

**Issue by flackr** [w3c/csswg-drafts#13997]
We frequently have bugs filed such as https://issues.chromium.org/516829766 where the rendering of a backdrop-filter or filter blur incurs a significant performance penalty. I believe css blurs are so expensive because they are calculated as a gaussian blur, https://drafts.csswg.org/filter-effects-1/#feGaussianBlurElement. In chrome, when you animate this blur we compile a new shader per blur radii visited in the animation which is why in https://issues.chromium.org/40831523 I suggested we may be able to produce imperfect intermediate frames while animating.

However, even if we can make it slightly better, I think in many use cases where developers apply blurs they do not care about the exact blur quality. I propose that we have some way for using simpler cheaper blurs both for static blurs and for animating blurs. E.g.

```css
filter-rendering: smooth | high-quality
```

The name, filter-rendering, and values smooth | high-quality, were chosen to align with [image-rendering](https://www.w3.org/TR/css-images-3/#the-image-rendering) as it has similar implications. This would allow developers to opt into making their blurs significantly less expensive to render.

E.g. a UA could choose to implement smooth filtering by using a low resolution rendering and using simple gpu scaling to smooth between those sampled pixels.

**Comment by flackr** [comment]
We could also consider adding a `pixelated` value to support pixelated blurs.

**Comment by jakearchibald** [comment]
A case where I would have benefited from this is when blurring low quality image previews. But also in that case, I don't want a filter blur, because of how it handles edge pixels, so I end up having to use backdrop-filter, which adds complexity to the solution.

It'd be nice to do something like `filter: blur(10px simple edge-repeat)` on an img. 

I realise there are complexities in calculating the edge, and handling pixels that are outside the edge if we use the element's bounding box.

Hmm, although GPU scaling may be too simple for low quality image previews.

**Comment by bramus** [comment]
> It'd be nice to do something like `filter: blur(10px simple edge-repeat)` on an img.

Adding an extra argument to `blur()` was my gut reaction here as well. But that would require authors to adjust all function calls + require them to duplicate the declarations for browsers with and without support. With `filter-rendering: …`, they can just declare it (on `:root`) without needing to worry about compat.

**Comment by ydaniv** [comment]
> > It'd be nice to do something like filter: blur(10px simple edge-repeat) on an img.
>
> Adding an extra argument to blur() was my gut reaction here as well. But that would require authors to adjust all function calls + require them to duplicate the declarations for browsers with and without support. With filter-rendering: …, they can just declare it (on :root) without needing to worry about compat.

Makes sense, but we also need to consider whether this scales well future-wise, i.e. does this property makes sense for other types of filters?
I think in case of `filter-rendering` it's only relevant for filters involving some sort of a convolution, so only `blur` and `drop-shadow` for now. But probably if we want to introduce more filters from SVG in the future it may become useful there as well.

Also, will that property also affect `backdrop-filter`? How about `filter()`?

**Comment by jakearchibald** [comment]
@bramus 

> With `filter-rendering: …`, they can just declare it (on `:root`) without needing to worry about compat.

Will developers really want to do that? If the blurs are going to be as low quality as @flackr suggests, then it seems like I'd only want to apply it in particular cases.

@ydaniv 

> but we also need to consider whether this scales well future-wise, i.e. does this property makes sense for other types of filters?

Strong agree. What does "smooth" mean in terms of `drop-shadow`? If you add a really high `contrast()`, it tends to produce a 'sharp' image - so what does "smooth" mean there? If we added an "outline" filter, what would "smooth" mean there?

Even if this is a separate property, I think it should only target `blur` on `filter`, and have a separate property for targeting `blur` on `backdrop-filter`. (or just make it part of `blur()`)

**Comment by jakearchibald** [comment]
https://random-stuff.jakearchibald.com/blurs/ - I slopped around a bit. I'm worried about the quality and animatability of "low resolution rendering and using simple gpu scaling", but maybe I'm missing a technique.

Something like Dual Kawase looks good, with cross-fading between the phases, as in the "Dual Kawase" demo there. Although it doesn't match CSS for px values in my demo.

**Comment by flackr** [comment]
Backdrop-filter is a very important use case given the prevalance of glass effects. I also expect this to be used in blur extensions like the linear-blur #13285. I'm open to defining it in the function parameters instead though.

> I slopped around a bit. I'm worried about the quality and animatability of "low resolution rendering and using simple gpu scaling", but maybe I'm missing a technique.

One thing to note, is that in order to sample the outside pixels the same way that blur does you need to include them in your downsampled image. Also, in order to avoid jumps as the blur radius moves between subsequent downsamplings we would make this smooth by cross-fading between the two. I believe this is the same way that games ensure that switching between mipmapped textures doesn't result in popping. Another option is to downsample partially, and apply a blur kernel over the downsampled result to get a combination of the two methods.

That said, my hope is to get agreement that we want to pursue the ability to have cheaper blurs even if we don't agree yet on which method to use.

**Comment by LeaVerou** [comment]
If this is primarily for animations, can't it be implemented as a UA optimization? Then it can be applied heuristically depending on the animation specifics, and all the spec needs to do is give permission for these types of optimizations.

@jakearchibald can you open another issue about this? I think it would be a great improvement, but is orthogonal to this.

**Comment by flackr** [comment]
> If this is primarily for animations, can't it be implemented as a UA optimization?

I mentioned this as a possible UA optimization while animating in the OP, but raised this issue because it applies not just for animations. It is quite common in modern UI's to apply a backdrop filter blur behind large overlay elements. The quality of this blur is usually not important since the focus is the foreground of the overlaying element, but the cost is significant.

**Comment by fantasai** [comment]
I agree with @ydaniv's concerns about how a filter-quality property would affect (or not affect) other filters. Adding a parameter to `blur()` or creating a variant `fast-blur()` makes more sense to me for this reason.

**Comment by yisibl** [comment]
I have a slight concern about the name `fast-blur()`: if, someday in the future, `blur()` becomes just as fast as `fast-blur()`, then the word “fast” in that name will no longer imply that it’s faster.

**Comment by fantasai** [comment]
Proposal A: [Add a filter rendering quality property](https://github.com/w3c/csswg-drafts/issues/13997#issue-4564579204).
Proposal B: [Add more parameters to the filter functions](https://github.com/w3c/csswg-drafts/issues/13997#issuecomment-5100392648).
Proposal C: [Add a variant filter function](https://github.com/w3c/csswg-drafts/issues/13997#issuecomment-5416848490).

**Comment by romainmenke** [comment]
A property could be used by multiple properties/filters on the same element but also makes it impossible to make different choices. You might want a fast background blur but a high quality shadow on the same element.
